### PR TITLE
Update Gradle to v7 so that can build with JDK 8 to 18

### PR DESCRIPTION
**Pull request by SingingBush**
There's a few changes here that are worth pulling in. I wanted to ensure this library works with JDK 17 & 18 so I updated the Github Action to perform build/tests with multiple JDK versions (8, 11, 17, 18). This required Gradle being updated to version 7 which in turn needed numerous plugins to be updated. As part of this I restructured the action so that the publish job is only done once tests have passed on each JDK version:
image

### CODE REVIEW

File: src/test/groovy/net/fortuna/ical4j/model/DateTimeSpec.groovy [hunk 1]
```diff
@@ -112,10 +112,10 @@ class DateTimeSpec extends Specification {
                 }
         ]
 
-        dateTimeString << ['20110327T000000', '20110327T000000']
+        dateTimeString << ['20110327T000000Z', '20110327T000000Z']
```
Comment (benfortuna) [inline src/test/groovy/net/fortuna/ical4j/model/DateTimeSpec.groovy:115]: Is this change required for the gradle upgrade? By using UTC format DATE-TIME i suspect it won't properly test the custom timezone..

File: src/test/resources/samples/invalid/eli_courtwright.ics [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
- riends Meeting of Washington on Humor in the Bible
-DESCRIPTION:The focus of the discussion will be on Biblical humor as int
- ended by the author.\n\nOur monthly potluck is the first Sunday of each 
- month at 6:30 PM. This is an opportunity for fellowship among Quakers a
- nd attenders. Bring a dish to share\; family members and friends are al
- ways welcome. Note: William Penn House is an alcohol and tobacco free f
- acility.
-END:VEVENT
-END:VCALENDAR
+BEGIN:VCALENDAR
```
Comment (benfortuna) [inline src/test/resources/samples/invalid/eli_courtwright.ics:1]: Is the change for the ICS test files relating to line endings? Not sure if this was intentional or not..

File: .github/workflows/gradle.yml [hunk 3]
```diff
@@ -6,12 +6,37 @@ permissions: read-all
 
 jobs:
   gradle:
+    name: Test with Java ${{ matrix.jdk }}
```
Comment (benfortuna) [inline .github/workflows/gradle.yml:9]: Awesome, thanks so much for doing this. I thought there may have been something that blocked the gradle v7 upgrade but if it is working then happy to accept this.

File: src/test/groovy/net/fortuna/ical4j/model/DateTimeSpec.groovy (hunk 1 above)
Comment (SingingBush) [reply src/test/groovy/net/fortuna/ical4j/model/DateTimeSpec.groovy:115]: not sure if it was change to Groovy version or a plugin but for some reason this test was failing after the changes and specifying UTC fixed the problem

File: src/test/resources/samples/invalid/eli_courtwright.ics (hunk 2 above)
Comment (SingingBush) [reply src/test/resources/samples/invalid/eli_courtwright.ics:1]: I'd thought it was only encoding. As soon as I pulled the repo a bunch of the ics files had shown changes and I was unable to revert or stash them them as it just kept happening automatically. I thought that perhaps Intellij was changing them to UTF-8 so I closed the IDE and tried reverting them in terminal but still the same. I ended up just committing them as I couldn't work around it.

File: .github/workflows/gradle.yml (hunk 3 above)
Comment (SingingBush) [reply .github/workflows/gradle.yml:9]: It certainly seemed ok once all the changes were in place. There are still warnings in the logs that perhaps need further investigating but I think that is fine to merge and look at them later. I'm using Java 17 everywhere now so am fairly keen to make sure the project builds on JDK 17, which is only supported on Gradle 7.

**Comment by SingingBush** [comment]
I'm happy to follow this up with another PR to address any knock on effects of these changes if required. I've used ical4j in the past and am hoping to use it in some upcoming work only this time round I'm targeting JDK 17 and using GraalVM to do native builds. For this reason getting fairly minimal dependencies is fairly important so I'm looking forward to #242 being complete.

File: .github/workflows/gradle.yml (hunk 3 above)
Comment (SingingBush) [reply .github/workflows/gradle.yml:9]: One thing worth pointing out is that Gradle doesn't do optional dependencies in the way that Maven does, instead it has the concept of feature variants: https://docs.gradle.org/current/userguide/feature_variants.html 

It's probably worth taking some time to use feature variants instead of marking dependencies as optional.

File: src/test/resources/samples/invalid/eli_courtwright.ics (hunk 2 above)
Comment (benfortuna) [reply src/test/resources/samples/invalid/eli_courtwright.ics:1]: It's weird because I am also using IntelliJ, but when I check out your fork it also applies changes to revert the line endings. 

Note that there is a `.gitattributes` file to force the `CRLF` line endings for *.ics files, but for some reason it doesn't seem to work for you I guess. If I may I might try to revert this in your fork to reduce the noise in the PR.
